iceoryx_utils/include/iceoryx_utils/internal/concurrent/lockfree_queue/cyclic_index.inl

```diff
@@ -81,5 +87,11 @@ bool CyclicIndex<CycleLength, ValueType>::isOneCycleBehind(const CyclicIndex& ot
     }
     return (thisCycle + 1 == otherCycle);
 }
+
+template <uint64_t CycleLength, typename ValueType>
+int64_t CyclicIndex<CycleLength, ValueType>::operator-(const CyclicIndex<CycleLength, ValueType>& rhs) const
+{
+    return m_value - rhs.m_value;
```

Comment:
should the underflow be handled, e.g. by returning 0?

Reply:
Interesting topic. I explicitly want to detect negative differences, see how it is used in popIfSizeIsAtLeast in index_queue.inl.
Theoretcially I always subtract something smaller (or equal), so the difference cannot be < 0. But due to the way atomics work I may actually subtract something larger and the I will detect it, to a degree (if the difference is smaller then 2^63 basically). This is similar to the ABA problem, only happens if your load is severely outdated in our use. But then we have other problems...

Furthermore, in isolation I would expect the operator to provide negative differences... The problem is: the underlying values are unsigned 64 bit, the difference is not necessarily representable in signed 64 bit. I do not want to use signed integers for the value though, the relevant differences for our use case are representable.

TLDR: truncating to zero is not really a useful option, and besides would incur a cost everytime (unles the hardware provides a truncated difference operation).

Reply:
No. As I wrote, we cannot do this here. The queue is robust against this kind of error from my point of view. Negative differences are no problem. It just becomes a problem if the values are to far apart and the difference is not representable. But this will only happen if one thread gets massively delayed and would take over 250 years assuming a push operation takes 1ns (it will take more). It is a similar argument as for the ABA counter.

Reply:
This is not a matter ofperformance, but of correctness. Truncating would be undesired behavior and lead to the wrong result.